### PR TITLE
DM-37501: deprecate synchronous callbacks

### PR DESCRIPTION
**Pull request by r-owen**

### CODE REVIEW

File: python/lsst/ts/salobj/csc_commander.py [hunk 1]
```diff
@@ -272,6 +285,20 @@ def __init__(
...
+        all_members = inspect.getmembers(self, predicate=inspect.ismethod)
+        sync_do_member_names = [
+            member[0]
+            for member in all_members
+            if member[0].startswith("do_")
+            and not inspect.iscoroutinefunction(member[1])
+        ]
+        if sync_do_member_names:
+            warnings.warn(
+                f"Synchronous do_ methods are deprecated: {sorted(sync_do_member_names)}",
```
Comment (pkubanek) [inline python/lsst/ts/salobj/csc_commander.py:298]: I have a slight preference for use of ','.join, as listing method names in an array (as [m1,m2,..] is a bit brutal.

File: python/lsst/ts/salobj/topics/read_topic.py [hunk 2]
```diff
@@ -428,6 +430,16 @@ def callback(
...
     @callback.setter
     def callback(self, func: None | CallbackType) -> None:
+        if func is not None:
+            if not callable(func):
+                raise TypeError(f"callback {func} not callable")
+            if not inspect.iscoroutinefunction(func):
+                # TODO DM-37502: modify this to raise (and update doc string)
+                # once we drop support for synchronous callback functions.
+                warnings.warn(
+                    "callback should be asynchronous", category=DeprecationWarning
```
Comment (pkubanek) [inline python/lsst/ts/salobj/topics/read_topic.py:440]: Can that produce which callback? f"callback {func} should.. perhaps?

**Review by pkubanek — APPROVED** [review]
Looks good, see suggestions for minor improvements.

File: python/lsst/ts/salobj/csc_commander.py (hunk 1 above)
Comment (r-owen) [reply python/lsst/ts/salobj/csc_commander.py:298]: I would prefer to keep this simple, and the [] act as a useful delimiter.

**Pull request merged**